cdap-watchdog/src/main/java/io/cdap/cdap/metrics/collect/Distribution.java

```diff
@@ -0,0 +1,136 @@
+     * long[8] and the 5th column is increased to 1. If bucket 53 needs to be incremented, the row already
+     * exists. So the 6th column is incremented.
+     *
+     * In the worst case, this can take storage for 64 longs + 8 row references. But in the normal case, the range
+     * of values for a metric will probably within a 256X range i.e. max value <= min * 256. So at most two rows
+     * will have to be initialized, resulting in 8*2 long number storage + 8 object references.
+     */
+    private final long[][] bucketCounts = new long[8][];
+
+    /**
+     * If count is zero then this field must be zero. Mean can be easily calculated from sum and count.
+     */
+    private double mean;
+
+    private long count;
+
+    private int numBuckets;
+
+    public void add(long value) {
+        int index = findBucketNumber(value);
+        // online calculation of mean and sumOfSquaredDeviation
+        count += 1;
+        double dev = value - mean;
+        mean += dev / count;
+
+        int i = getRow(index);
+        int j = getCol(index);
+        if (bucketCounts[i] == null) {
+            bucketCounts[i] = new long[8];
+        }
+        if (bucketCounts[i][j] == 0) {
+            numBuckets++;
+        }
+        bucketCounts[i][j] += 1;
+    }
+
+    private int findBucketNumber(long value) {
+        if (value < 0) {
+            return 0;
+        }
+        // Algo to find bucket has logarithmic complexity. Max number of
+        // iterations is log_2 64. Note that it is logarithmic wrt number of
+        // buckets and not the value
+        int left = 0;
+        int right = MetricValue.NUM_FINITE_BUCKETS;
+        int mid;
+        while (left != right) {
+            mid = (left + right) / 2;
+            if (value >> mid == 0) {
+                right = mid;
+            } else {
+                left = mid + 1;
+            }
+        }
+        return left + 1;
+    }
+
+    private int getRow(int index) {
+        // index/8
+        return index >> 3;
+    }
+
+    private int getCol(int index) {
+        // index%8
+        return index & 7;
+    }
+
+    public MetricValue getMetricValue(String metricName) {
+        long bucketMask = 0;
+        int i , j;
+        long[] bucketCountsArr = new long[numBuckets];
+        int bucketCountIndex = 0;
+        for (int index = 0; index < 64 && bucketCountIndex < numBuckets; index++) {
+            i = getRow(index);
+            j = getCol(index);
+            if (bucketCounts[i] != null && bucketCounts[i][j] > 0) {
+                bucketMask |= 1L << index;
+                bucketCountsArr[bucketCountIndex] = bucketCounts[i][j];
+                bucketCountIndex++;
+            }
+        }
+        return new MetricValue(metricName, bucketCountsArr, bucketMask, mean);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append((String.format("count: %d, numBuckets:%d, mean:%32.2f %n",
```

Comment:
I don't think you need double brackets here. Also please be consistent with spacing and other toString implementations in cdap. toString should usually indicate what is it (Distribution), have surrounding curly brackets.
Why are you adding line separators? Objects toString is usually embedded within a phrase / log line and newlines can break the message, so don't add unless really needed

Reply:
Removed the double brackets, fixed spacing etc. The new lines make the string representation more readable. Ex.
Distribution {count: 9571620, numBuckets: 1, mean:                            1.00 
0: [0, 0, 9571620, 0, 0, 0, 0, 0]
1: null
2: null
3: null
4: null
5: null
6: null
7: null
}